File: implementations/go/s57.go
package b57

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/binary"
	"time"

	"lukechampine.com/blake3"
)

const (
	S57Version byte = 0x01
)

var s57EntropyReader = s57ReadEntropyBytes

// S57Keys holds derived domain-separated keys.
type S57Keys struct {
	B57AES256Key []byte
	H57Key       []byte
	ID57Key      []byte
	KeyID        byte
}

// S57Config configures S57 secure composition.
type S57Config struct {
	ServerSecretKey []byte
	EnvironmentSalt []byte
	KeyID           byte
}

// S57 provides secure composition APIs over B57/H57/ID57/R57.
type S57 struct {
	serverSecret []byte
	envSalt      []byte
	keys         S57Keys
}

// NewS57 constructs an S57 instance and derives domain-separated keys.
func NewS57(config S57Config) (*S57, error) {
	if len(config.ServerSecretKey) < 32 {
		return nil, NewKeyInvalidError()
	}
	if len(config.EnvironmentSalt) == 0 {
		return nil, NewKeyInvalidError()
	}

	s := &S57{
		serverSecret: append([]byte(nil), config.ServerSecretKey...),
		envSalt:      append([]byte(nil), config.EnvironmentSalt...),
	}
	if config.KeyID == 0 {
		config.KeyID = 1
	}

	keys, err := s.ResolveKeys(config.KeyID)
	if err != nil {
		return nil, err
	}
	s.keys = keys
	return s, nil
}

// ResolveKeys derives B57_AES_256_KEY, H57_KEY and ID57_KEY.
func (s *S57) ResolveKeys(keyID byte) (S57Keys, error) {
	seed := append(append([]byte(nil), s.serverSecret...), s.envSalt...)
	if len(seed) < 32 {
		return S57Keys{}, NewKeyInvalidError()
	}

	b57AES := make([]byte, 32)
	h57Key := make([]byte, 32)
	id57Key := make([]byte, 32)

	blake3.DeriveKey(b57AES, "S57/B57_AES_256/v1", seed)
	blake3.DeriveKey(h57Key, "S57/H57_KEY/v1", seed)
	blake3.DeriveKey(id57Key, "S57/ID57_KEY/v1", seed)

	return S57Keys{
		B57AES256Key: b57AES,
		H57Key:       h57Key,
		ID57Key:      id57Key,
		KeyID:        keyID,
	}, nil
}

// Hash computes keyed H57 with S57 profile lengths (128, 256, 512).
func (s *S57) Hash(data []byte, length H57Length) (string, error) {
	bits, err := s.resolveS57H57Bits(length)
	if err != nil {
		return "", err
	}

	requested := (bits + 7) / 8
	digest := blake3.New(requested, s.keys.H57Key)
	_, _ = digest.Write(data)
	out := digest.Sum(nil)
	return Encode(out), nil
}

// ID computes keyed ID57 with S57 profile lengths (128, 256, 512).
func (s *S57) ID(data []byte, length ID57Length) (string, error) {
	b, ok := id57BitsByLength[length]
	if !ok {
		if length == ID57Default {
			b = 128
		} else {
			return "", NewInvalidLengthEnumError(int(length))
		}
	}
	if b != 128 && b != 256 && b != 512 {
		return "", NewInvalidLengthEnumError(int(length))
	}

	requested := (b + 7) / 8
	digest := blake3.New(requested, s.keys.ID57Key)
	_, _ = digest.Write(data)
	out := digest.Sum(nil)
	maskExcessBits(out, b)
	return Encode(out), nil
}

// Random returns CSPRNG-based 128-bit random identifier.
func (s *S57) Random() (string, error) {
	raw, err := s57EntropyReader(16)
	if err != nil {
		return "", NewInsufficientEntropyError()
	}
	return encodeR57128(raw), nil
}

// RandomTime mixes coarse timestamp with randomness.
func (s *S57) RandomTime() (string, error) {
	r, err := s57EntropyReader(12)
	if err != nil {
		return "", NewInsufficientEntropyError()
	}
	t := make([]byte, 4)
	binary.BigEndian.PutUint32(t, uint32(time.Now().Unix()))
	m := blake3.Sum256(append(t, r...))
	return encodeR57128(m[:16]), nil
}

// RandomCounter mixes user counter with randomness.
func (s *S57) RandomCounter(counter uint32) (string, error) {
	r, err := s57EntropyReader(12)
	if err != nil {
		return "", NewInsufficientEntropyError()
	}
	c := make([]byte, 4)
	binary.BigEndian.PutUint32(c, counter)
	m := blake3.Sum256(append(c, r...))
	return encodeR57128(m[:16]), nil
}

// RandomSession uses keyed derivation based on session secret and nonce.
func (s *S57) RandomSession(sessionSecret []byte) (string, error) {
	nonce, err := s57EntropyReader(16)
	if err != nil {
		return "", NewInsufficientEntropyError()
	}
	key := make([]byte, 32)
	blake3.DeriveKey(key, "S57/R57_SESSION/v1", sessionSecret)
	h := blake3.New(16, key)
	_, _ = h.Write(nonce)
	return encodeR57128(h.Sum(nil)), nil
}

// RandomDevice uses keyed derivation based on device secret and nonce.
func (s *S57) RandomDevice(deviceSecret []byte) (string, error) {
	nonce, err := s57EntropyReader(16)
	if err != nil {
		return "", NewInsufficientEntropyError()
	}
	key := make([]byte, 32)
	blake3.DeriveKey(key, "S57/R57_DEVICE/v1", deviceSecret)
	h := blake3.New(16, key)
	_, _ = h.Write(nonce)
	return encodeR57128(h.Sum(nil)), nil
}

// RandomDerived is deterministic for identical (masterSecret, uniqueInput).
func (s *S57) RandomDerived(masterSecret, uniqueInput []byte) (string, error) {
	key := make([]byte, 32)
	blake3.DeriveKey(key, "S57/R57_DERIVED/v1", masterSecret)
	h := blake3.New(16, key)
	_, _ = h.Write(uniqueInput)
	return encodeR57128(h.Sum(nil)), nil
}

// RandomHardened re-hashes CSPRNG output.
func (s *S57) RandomHardened() (string, error) {
	raw, err := s57EntropyReader(16)
	if err != nil {
		return "", NewInsufficientEntropyError()
	}
	m := blake3.Sum256(raw)
	return encodeR57128(m[:16]), nil
}

// RandomHybrid mixes one or more entropy sources.
func (s *S57) RandomHybrid(sources ...[]byte) (string, error) {
	if len(sources) == 0 {
		return "", NewInvalidInputError("random_hybrid requires at least one source")
	}
	total := 0
	for _, src := range sources {
		total += len(src)
	}
	buf := make([]byte, 0, total)
	for _, src := range sources {
		buf = append(buf, src...)
	}
	m := blake3.Sum256(buf)
	return encodeR57128(m[:16]), nil
}

// Encrypt encrypts plaintext with AES-256-GCM envelope encoded in B57.
func (s *S57) Encrypt(plaintext, aad []byte) (string, error) {
	if len(s.keys.B57AES256Key) != 32 {
		return "", NewKeyInvalidError()
	}
	nonce, err := s57EntropyReader(12)
	if err != nil {
		return "", NewInsufficientEntropyError()
	}

	block, err := aes.NewCipher(s.keys.B57AES256Key)
	if err != nil {
		return "", err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return "", err
	}

	ciphertextWithTag := gcm.Seal(nil, nonce, plaintext, aad)
	env := make([]byte, 0, 2+len(nonce)+len(ciphertextWithTag))
	env = append(env, S57Version)
	env = append(env, s.keys.KeyID)
	env = append(env, nonce...)
	env = append(env, ciphertextWithTag...)

	return Encode(env), nil
}

// Decrypt decrypts a B57-encoded AES-256-GCM envelope.
func (s *S57) Decrypt(b57String string, aad []byte, expectedKeyID ...byte) ([]byte, error) {
	env, err := Decode(b57String)
	if err != nil {
		return nil, err
	}
	if len(env) < 2+12+16 {
		return nil, NewInvalidInputError("invalid envelope length")
	}

	version := env[0]
	if version != S57Version {
		return nil, NewInvalidVersionError(version)
	}

	keyID := env[1]
	if len(expectedKeyID) > 0 && keyID != expectedKeyID[0] {
		return nil, NewKeyUnavailableError(keyID)
	}
	if keyID != s.keys.KeyID {
		return nil, NewKeyUnavailableError(keyID)
	}

	if len(s.keys.B57AES256Key) != 32 {
		return nil, NewKeyUnavailableError(keyID)
	}

	nonce := env[2:14]
	ciphertextWithTag := env[14:]

	block, err := aes.NewCipher(s.keys.B57AES256Key)
	if err != nil {
		return nil, err
	}
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, err
	}

	pt, err := gcm.Open(nil, nonce, ciphertextWithTag, aad)
	if err != nil {
		return nil, NewAuthFailureError()
	}
	return pt, nil
}

func (s *S57) resolveS57H57Bits(length H57Length) (int, error) {
	if length == H57HashAuto {
		return 256, nil
	}
	b, ok := h57BitsByLength[length]
	if !ok {
		return 0, NewInvalidLengthEnumError(int(length))
	}
	if b != 128 && b != 256 && b != 512 {
		return 0, NewInvalidLengthEnumError(int(length))
	}
	return b, nil
}

func s57ReadEntropyBytes(length int) ([]byte, error) {
	buf := make([]byte, length)
	_, err := rand.Read(buf)
	if err != nil {
		return nil, err
	}
	return buf, nil
}

File: implementations/python/src/f57/s57.py
"""S57 secure composition profile."""

from __future__ import annotations

import secrets
import struct
import time
from dataclasses import dataclass
from typing import Optional

import blake3 as blake3_module
from cryptography.hazmat.primitives.ciphers.aead import AESGCM

from .b57 import encode, decode, is_valid, is_canonical
from .errors import (
    InvalidLengthEnumError,
    InvalidInputError,
    InvalidVersionError,
    AuthFailureError,
    KeyInvalidError,
    KeyUnavailableError,
)
from .h57 import H57Length
from .id57 import ID57Length

S57_VERSION = 0x01


@dataclass(frozen=True)
class S57Config:
    server_secret_key: bytes
    environment_salt: bytes
    key_id: int = 1


@dataclass(frozen=True)
class S57Keys:
    b57_aes_256_key: bytes
    h57_key: bytes
    id57_key: bytes
    key_id: int


class S57:
    def __init__(self, config: S57Config):
        if len(config.server_secret_key) < 32 or len(config.environment_salt) == 0:
            raise KeyInvalidError()

        self._server_secret = bytes(config.server_secret_key)
        self._env_salt = bytes(config.environment_salt)
        self._key_id = (config.key_id or 1) & 0xFF
        self.keys = self.resolve_keys(self._key_id)

    def resolve_keys(self, key_id: Optional[int] = None) -> S57Keys:
        effective = self._key_id if key_id is None else (key_id & 0xFF)
        seed = self._server_secret + self._env_salt
        if len(seed) < 32:
            raise KeyInvalidError()

        return S57Keys(
            b57_aes_256_key=_blake3_derive_key("S57/B57_AES_256/v1", seed, 32),
            h57_key=_blake3_derive_key("S57/H57_KEY/v1", seed, 32),
            id57_key=_blake3_derive_key("S57/ID57_KEY/v1", seed, 32),
            key_id=effective,
        )

    def hash(self, data: bytes, length: H57Length = H57Length.LEN256) -> str:
        bits = _resolve_s57_h57_bits(length)
        requested = (bits + 7) // 8
        out = _blake3_keyed(self.keys.h57_key, data, requested)
        out = bytearray(out)
        _mask_excess_bits(out, bits)
        return encode(bytes(out))

    def id(self, data: bytes, length: ID57Length = ID57Length.DEFAULT) -> str:
        effective = ID57Length.LEN128 if length == ID57Length.DEFAULT else length
        bits = _id57_bits(effective)
        if bits not in (128, 256, 512):
            raise InvalidLengthEnumError(length.value)

        requested = (bits + 7) // 8
        out = _blake3_keyed(self.keys.id57_key, data, requested)
        out = bytearray(out)
        _mask_excess_bits(out, bits)
        return encode(bytes(out))

    def random(self) -> str:
        return _encode_r57_128(secrets.token_bytes(16))

    def random_time(self) -> str:
        now_secs = int(time.time()) & 0xFFFFFFFF
        mixed = blake3_module.blake3(struct.pack(">I", now_secs) + secrets.token_bytes(12)).digest(length=16)
        return _encode_r57_128(mixed)

    def random_counter(self, counter: int) -> str:
        mixed = blake3_module.blake3(struct.pack(">I", counter & 0xFFFFFFFF) + secrets.token_bytes(12)).digest(length=16)
        return _encode_r57_128(mixed)

    def random_session(self, session_secret: bytes) -> str:
        key = _blake3_derive_key("S57/R57_SESSION/v1", bytes(session_secret), 32)
        mixed = _blake3_keyed(key, secrets.token_bytes(16), 16)
        return _encode_r57_128(mixed)

    def random_device(self, device_secret: bytes) -> str:
        key = _blake3_derive_key("S57/R57_DEVICE/v1", bytes(device_secret), 32)
        mixed = _blake3_keyed(key, secrets.token_bytes(16), 16)
        return _encode_r57_128(mixed)

    def random_derived(self, master_secret: bytes, unique_input: bytes) -> str:
        key = _blake3_derive_key("S57/R57_DERIVED/v1", bytes(master_secret), 32)
        mixed = _blake3_keyed(key, bytes(unique_input), 16)
        return _encode_r57_128(mixed)

    def random_hardened(self) -> str:
        mixed = blake3_module.blake3(secrets.token_bytes(16)).digest(length=16)
        return _encode_r57_128(mixed)

    def random_hybrid(self, *sources: bytes) -> str:
        if len(sources) == 0:
            raise InvalidInputError("random_hybrid requires at least one source")
        merged = b"".join(bytes(s) for s in sources)
        mixed = blake3_module.blake3(merged).digest(length=16)
        return _encode_r57_128(mixed)

    def encrypt(self, plaintext: bytes, aad: bytes = b"") -> str:
        if len(self.keys.b57_aes_256_key) != 32:
            raise KeyInvalidError()
        nonce = secrets.token_bytes(12)
        aesgcm = AESGCM(self.keys.b57_aes_256_key)
        ciphertext_with_tag = aesgcm.encrypt(nonce, bytes(plaintext), bytes(aad))

        envelope = bytes([S57_VERSION, self.keys.key_id]) + nonce + ciphertext_with_tag
        return encode(envelope)

    def decrypt(self, b57_string: str, aad: bytes = b"", expected_key_id: Optional[int] = None) -> bytes:
        env = decode(b57_string)
        if len(env) < 30:
            raise InvalidInputError("invalid envelope length")

        version = env[0]
        if version != S57_VERSION:
            raise InvalidVersionError(version)

        key_id = env[1]
        if expected_key_id is not None and key_id != (expected_key_id & 0xFF):
            raise KeyUnavailableError(key_id)
        if key_id != self.keys.key_id:
            raise KeyUnavailableError(key_id)

        nonce = env[2:14]
        ciphertext_with_tag = env[14:]
        if len(ciphertext_with_tag) < 16:
            raise InvalidInputError("invalid envelope length")

        aesgcm = AESGCM(self.keys.b57_aes_256_key)
        try:
            return aesgcm.decrypt(nonce, ciphertext_with_tag, bytes(aad))
        except Exception as exc:
            raise AuthFailureError() from exc


def s57_is_valid(value: str) -> bool:
    return is_valid(value)


def s57_is_canonical(value: str) -> bool:
    return is_canonical(value)


def _resolve_s57_h57_bits(length: H57Length) -> int:
    if length == H57Length.HASH_AUTO:
        return 256
    if length == H57Length.LEN128:
        return 128
    if length == H57Length.LEN256:
        return 256
    if length == H57Length.LEN512:
        return 512
    raise InvalidLengthEnumError(length.value)


def _id57_bits(length: ID57Length) -> int:
    bits = {
        ID57Length.LEN8: 8,
        ID57Length.LEN16: 16,
        ID57Length.LEN23: 23,
        ID57Length.LEN29: 29,
        ID57Length.LEN32: 32,
        ID57Length.LEN47: 47,
        ID57Length.LEN64: 64,
        ID57Length.LEN70: 70,
        ID57Length.LEN93: 93,
        ID57Length.LEN128: 128,
        ID57Length.LEN186: 186,
        ID57Length.LEN256: 256,
        ID57Length.LEN373: 373,
        ID57Length.LEN512: 512,
    }
    if length not in bits:
        raise InvalidLengthEnumError(length.value)
    return bits[length]


def _blake3_derive_key(context: str, seed: bytes, out_len: int) -> bytes:
    h = blake3_module.blake3(derive_key_context=context)
    h.update(seed)
    return h.digest(length=out_len)


def _blake3_keyed(key: bytes, data: bytes, out_len: int) -> bytes:
    h = blake3_module.blake3(key=key)
    h.update(data)
    return h.digest(length=out_len)


def _mask_excess_bits(data: bytearray, bit_length: int) -> None:
    if not data:
        return
    excess = len(data) * 8 - bit_length
    if excess <= 0:
        return
    mask = (0xFF << excess) & 0xFF
    data[-1] &= mask


def _encode_r57_128(raw: bytes) -> str:
    encoded = encode(raw)
    if len(encoded) == 22:
        return encoded

    derived = raw
    while len(encoded) < 22:
        derived = blake3_module.blake3(derived + b"\x57").digest(length=16)
        encoded = encode(derived)
    return encoded
